toolchain/semantics/semantics_context.cpp

```diff
@@ -380,6 +344,33 @@ auto SemanticsContext::CanonicalizeType(SemanticsNodeId node_id)
   return type_id;
 }
 
+auto SemanticsContext::CanonicalizeStructType(ParseTree::Node parse_node,
+                                              SemanticsNodeBlockId refs_id)
+    -> SemanticsTypeId {
+  // Construct the field structure for lookup.
+  auto refs = semantics_->GetNodeBlock(refs_id);
+  llvm::SmallVector<std::pair<SemanticsStringId, SemanticsTypeId>> fields;
+  fields.reserve(refs.size());
+  for (const auto& ref_id : refs) {
+    auto ref = semantics_->GetNode(ref_id);
+    fields.push_back({ref.GetAsStructTypeField(), ref.type_id()});
+  }
+
+  // If a struct with matching fields was already created, reuse it.
+  auto it = canonical_struct_types_.find(fields);
+  if (it != canonical_struct_types_.end()) {
+    return it->second;
+  }
```

Comment:
If the `FoldingSet` approach doesn't work out (see below, or above... not sure which order my comments show up here), can you leave a TODO to revisit this for efficiency? Not worth doing it right away, but this seems likely to get into a loop doing heap allocation just to query the canonicalization. Pretty sure we can do better by incrementally computing the hash/key while walking the semantics rather than allocating it, or doing some other somewhat clever thing here. But seems much better to think about this in the future rather than now.

Reply:
Switched to FoldingSet.